loopy/transform/precompute.py

```diff
@@ -1074,10 +1075,14 @@ def add_assumptions(d):
         elif temp_var.dtype is not lp.auto and dtype is lp.auto:
             dtype = temp_var.dtype
         elif temp_var.dtype is not lp.auto and dtype is not lp.auto:
-            if temp_var.dtype != dtype:
+            if isinstance(temp_var.dtype, LoopyType):
+                temp_var_dtype = temp_var.dtype.numpy_dtype
```

Comment:
IIUC this seems to work in the opposite way from what I'd expect. `TemporaryVariable.dtype` *should* be a `LoopyType`. By now we even have a type annotation that says so:

https://github.com/inducer/loopy/blob/5d83454a0a50e9c1e8d93c934ec27989802d6a9d/loopy/kernel/array.py#L703

Reply:
Aah I see. How does https://github.com/inducer/loopy/compare/4cddc9bc46f57a0e67ed3163d91b57ff82c016f4..aa77cf0d36106b9d1527480750220add016ea793 look?